bootloader: fix splash screen shutdown issues

rokm opened the pull request:
This PR starts by adding bunch of new tests for splash screen - in particular, its *shutdown*. The one test that we had up until now is, in fact, "interactive test", where we expect the application to be alive for at least the specified amount of time (10 seconds), before we *terminate* it. Would the frozen application shut down without crashing? Would the onefile build leak its temporary directory? [Those who know less sleep better...](https://github.com/rokm/pyinstaller/actions/runs/9384136834/job/25839289530).

We then instrument the bootloader's cleanup code with couple of additional debug messages in an attempt to pin-point the issues that the new tests flushed out.

Turns out that in splash screen cleanup code, we are unloading Tcl shared library before Tk one. Reverting the order seems to deal with crashes in the new tests. However, the `DllMain` function in Tk seems to contain provisions for Tcl being unloaded before Tk; although on the other hand, the surrounding comments imply that this is considered an abnormal situation... Either way, the correct order seems to improve the situation, so let's go with it. And keep an eye on CI to see if we have any sporadic failures of those tests.

The above seems to take care of all issues but one - now the onefile variant of new tests seems to be consistently failing on test-msys2 (mingw32, i686). The frozen application is failing to clean up its temporary directory, and with the recently-refactored bootloader, this now triggers an error on our CI (as per `PYINSTALLER_STRICT_UNPACK_MODE` environment variable).

Again, we add couple of debug messages to make debugging such problems easier, and we learn that `libgcc_s_dw2-1.dll` and `libwinpthread-1.dll` cannot be removed. Those are dependencies of `tcl86.dll`, so we are dealing with a variant of #7106 - loading Tcl and Tk DLLs for splash screen loads additional DLLs into the process, but when we unload Tk and Tcl, those additional DLLs are not unloaded, and thus they become un-removable.

So as a work-around, if we fail to remove the temporary directory, we now enumerate all loaded DLLs, look up their file name to identify the ones that originate from the application's temporary directory, and attempt to force-unload them by repeatedly calling `FreeLibrary` until the call fails. And then try removing the temporary directory again. As questionable and hacky as this sounds, it does seem to do the trick; both for msys2/mingw32 and for UPX-processed Tcl/Tk DLLs - so we can close #7106.

rokm committed:
tests: move splash screen test into separate file

Move the splash screen test into separate `test_splash.py` file.
This way, when we add new splash screen tests, they will be all
in the same place.

Rename the test into `test_splash_screen_running`.

rokm committed:
tests: add tests for shut-down of splash-screen-enabled program

Add tests for ensuring that splash-enabled program properly exits
(i.e., there are no crashes in bootloader during shutdown). This
is not covered by the existing interactive test, because in that
one, we terminate the program after it runs for the expected amount
of time.

rokm committed:
bootloader: add extra debug messages for easier debugging

Add an `LOADER: end of process reached!` message at the end of
both `_pyi_main_onedir_or_onefile_child` and
`_pyi_main_onefile_parent`; this allows us to quickly determine
whether the process finished successfully or crashed during
cleanup.

Add couple of debug messages to `pyi_splash_finalize` to pinpoint
the issues with splash screen shutdown.

rokm committed:
bootloader: splash: unload Tk shared library before Tcl one

Switch the order in which we unload Tk and Tcl shared libraries;
Tk should be unloaded before Tcl.

The DllMain in Tk seems to contain provisions for Tcl being
unloaded before Tk, but we should get our order correct nevertheless.

Fixing the unload order seems to sort out crashes during splash
screen clean up in Windows, as demonstrated by newly-added
tests.

rokm committed:
bootloader: emit debug messages for file/directories we fail to remove

During onefile's temporary directory cleanup, emit debug messages
for files and directories we fail to remove, to simplify debugging.

Keep the messages restricted to debug in order to avoid spamming
error and warning dialogs in noconsole/windowed mode on Windows.
(The failure to remove the application's top-level temporary
directory itself already results in a warning/error message).

rokm committed:
bootloader: implement last-ditch attempt to force-unload bundled DLLs

On Windows, if we fail to remove onefile application's temporary
directory, iterate over all DLLs loaded by the process, identify
the ones that originate from application's temporary directory,
and attempt to force-unload them. Then try to remove the temporary
directory again.

This is the last-ditch attempt at working around cases when Tcl/Tk
DLLs that are loaded for splash screen pull in additional dependencies,
but fail to unload them when they are unloaded themselves. As they
remain loaded in the process, such DLLs prevent removal of the
application's temporary directory.

Known examples are `tcl86.dll` from `mingw-w64-i686-tcl` 8.6.12-3
(32-bit msys2/mingw32 environment), which leak `libgcc_s_dw2-1.dll`
and `libwinpthread-1.dll`, and 64-bit Tcl/Tk DLLs processes with
UPX, which leak `VCRUNTIME140.dll`.

rokm commented:
That failed linux pipeline on the first run confirms my suspicion that there is also a race condition involved, and I think I'm finally starting to see what is going on...

rokm committed:
bootloader: fix potential race condition during splash screen cleanup

Move the cleanup part that should be performed in the splash screen
thread (i.e., cleanup of Tcl interpreter instance, stored in
`splash->interp`) from `pyi_splash_finalize` into the cleanup
codepath of `_splash_init` (instead of having the latter call
`pyi_splash_finalize`).

This greatly simplifies `pyi_splash_finalize` implementation,
but also makes it easier to spot a potential race condition: if
frozen application closes splash screen just before it shuts
down, the splash screen thread might start its cleanup, and
reach the part where it cleans up Tcl interpreter instance.
At the same time, the main thread might call `pyi_splash_finalize`
as part of its own cleanup; so it might happen that
`pyi_splash_finalize` sees `splash->interp` being `NULL` and
proceeds with unloading of Tk and Tcl shared library, while the
splash screen thread is still running and using Tcl/Tk functions
to manipulate mutexes and wait condition.

Add another mutex to ensure synchronization between
`pyi_splash_finalize` and cleanup codepath in `_splash_init`.

rokm committed:
bootloader: splash: finalize mutexes during cleanup

Finalize the mutexes used by splash screen at the end of splash
screen resources cleanup.

bwoodsend commented on tests/functional/test_splash.py:
```diff
@@ -40,3 +42,55 @@ def test_splash_screen_running(pyi_builder_spec, capfd, monkeypatch, build_mode,
+# - splash screen is kept running until the end
+# - user imports pyi_splash, which calls pyi_splash.close() via atexit()
+# - user imports pyi_splash and calls pyi_splash.close()
+@pytest.mark.skipif(is_darwin, reason="Splash screen is not supported on macOS.")
+@pytest.mark.xfail(is_musl, reason="musl + tkinter is known to cause mysterious segfaults.")
```
Could we scope these to the whole file? i.e. Global variable `pytestmark = [pytest.mark.skipif(is_darwin, ...), pytest.xfail(is_musl, ...)]`

rokm replied:
Yeah, good idea.

rokm commented:
I rue the day I decided to pick this fight...

rokm committed:
bootloader: splash: another attempt at fixing race condition in shutdown

Consolidate the `exit_mutex` and `cleanup_mutex`.

bwoodsend commented:
I wonder if it might have been easier just to use native APIs rather than tcl for this splash screen.

rokm committed:
bootloader: splash: make splash screen thread joinable

Make splash screen thread joinable, and wait for its exit by
joining it before calling Tcl_Finalize and unloading the Tk and
Tcl shared library.

rokm commented:
> I wonder if it might have been easier just to use native APIs rather than tcl for this splash screen.

Hmmm, perhaps. It would certainly avoid the issues with Tcl/Tk DLLs remaining loaded in the parent process and consequently not being removed on Windows. Although I'd still like to believe that those are caused by our shutdown code, and can be fixed...

rokm committed:
bootloader: splash: Tcl on Windows does not support joinable threads

At the time of writing, Tcl on Windows does not support joinable
threads, so use them only on non-Windows.
